dist/ExtUtils-ParseXS/lib/perlxs.pod

```diff
@@ -1,2065 +1,4706 @@
+first. For example, for a tied variable, this C-level get function will
+call the perl-level C<FETCH()> method and assign the return value of that
+to the SV. Only then will C<SvIV()> do its C<SvIOK()> check.
+
+When presented with an unknown SV, it should always have its magic checked
+before examining the values of the SVs flags.
+
+In total, the C<SvIV(sv)> macro does roughly the equivalent of:
+
+    if (SvGMAGICAL(sv))
+        mg_get(sv);   /* do FETCH() etc; update the SV's value / flags */
+    if (!SvIOK(sv))
+        sv_2iv(sv);   /* convert undef to 0, "1" to 1 etc */
+    return SvIVX(sv); /* use the raw value */
+
+You will see soon that XS's typemap templates mostly use high-level macros
+like C<SvIV()>, so this is usually all handled automatically for you. Only
+if you start to do your own type conversions will you need to worry about
+these details.
+
+Forgetting to test for, and to call, C<get> magic will typically appear to
+work fine until the first time someone passes a tied variable or similar
+to your XSUB, and C<FETCH()> doesn't get called. Accessing fields with
+C<SvPVX()> etc without testing for C<SvPOK()> first may access a field in
+a body which doesn't exist and possibly trigger a SEGV.
+
+Magic should only be called once per "use"; for example if a tied scalar
+is passed as an argument to your XSUB, you would expect C<FETCH()> to only
+be called once. Normally this is easy because you (or the typemap code)
+does a single C<SvIV()> call. Occasionally you may have explicitly called
+C<mg_get()> first, perhaps in order to check some flags; if so, you can
+skip a second magic call with variants like C<SvIV_nomg()>. For example:
+
+    SvGETMAGIC(sv); /* this calls mg_get() if SvGMAGICAL() */
+    if (SvNOK(sv))
+        /* special-case: do something with a floating-point value */
+    else {
+        IV i = SvIV_nomg(sv);
+        /* fall-back to treating it as an integer value */
+    }
+
+
+A Perl reference is just another type of scalar. It is indicated by
+C<SvROK()> being true, and the pointer to the referent SV is accessed
+using C<SvRV()>.
+
+The equivalent of C<SvIV()> for strings is
+
+    STRLEN len;
+    char *pv = SvPV(sv, len);
+
+(and variants) which both retrieves a string pointer and sets C<len> to
+its length. Note that there is no guarantee that after this call
+C<SvPOK(sv)> is true, nor that C<pv == SvPVX(sv)>. For example, C<sv> may
+be a reference to a blessed object with an overloaded stringify (C<"">)
+method.  In which case, behind the scenes there may be a temporary SV
+containing the result of the call to the method, with C<pv> pointing to
+I<that> SV's string buffer. C<sv> remains a reference. Similarly, a
+non-overloaded reference to an array may return a temporary string like
+C<"ARRAY(0x12345678)">.
+
+If you need to coerce an SV to a string (e.g. before modifying its string
+value) then use C<SvPV_force()> or one of its variants. For example if
```

Comment:
Perhaps "before directly modifying its string buffer".

You don't need to force normal before calling sv_setpvn().